src/Microsoft.Diagnostics.Monitoring.EventPipe/EventPipeStreamProvider.cs

```diff
@@ -28,7 +30,43 @@ public async Task<Stream> ProcessEvents(DiagnosticsClient client, TimeSpan durat
             EventPipeSession session = null;
             try
             {
-                session = await client.StartEventPipeSessionAsync(_sourceConfig.GetProviders(), _sourceConfig.RequestRundown, _sourceConfig.BufferSizeInMB, cancellationToken).ConfigureAwait(false);
+                IEnumerable<EventPipeProvider> providers = _sourceConfig.GetProviders();
+                int bufferSizeInMB = _sourceConfig.BufferSizeInMB;
+                long rundownKeyword = _sourceConfig.RundownKeyword;
+                RetryStrategy retryStrategy = _sourceConfig.RetryStrategy;
+                bool retry = true;
+                while (retry)
+                {
+                    retry = false;
+                    try
+                    {
+                        EventPipeSessionConfiguration config = new(providers, bufferSizeInMB, (rundownKeyword != 0), true, rundownKeyword);
+                        session = await client.StartEventPipeSessionAsync(config, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (UnsupportedCommandException e)
+                    {
+                        if (retryStrategy == RetryStrategy.DropKeywordKeepRundown)
```

Comment:
Can these be converted to `when` clauses so that the exception isn't handled when the condition is not met and then you don't have to rethrow the exception?